sql-plugin/src/main/scala/org/apache/spark/rapids/hybrid/CoalesceConvertIterator.scala

```diff
@@ -128,38 +128,68 @@ class CoalesceConvertIterator(cpuScanIter: Iterator[ColumnarBatch],
 }
 
 object CoalesceConvertIterator extends Logging {
-
-  def hostToDevice(hostIter: Iterator[Array[RapidsHostColumn]],
+  /**
+   * Consumes the RapidsHostBatchProducer and converts the HostColumnVectors to Device ones.
+   */
+  def hostToDevice(hostProducer: RapidsHostBatchProducer,
                    outputAttr: Seq[Attribute],
                    metrics: Map[String, GpuMetric]): Iterator[ColumnarBatch] = {
-    val dataTypes = outputAttr.map(_.dataType).toArray
-
-    hostIter.map { hostVectors =>
-      Option(TaskContext.get()).foreach { ctx =>
-        withResource(new NvtxWithMetrics("gpuAcquireC2C", NvtxColor.GREEN,
-          metrics("GpuAcquireTime"))) { _ =>
-          GpuSemaphore.acquireIfNecessary(ctx)
+    new Iterator[ColumnarBatch] {
+
+      private val dataTypes = outputAttr.map(_.dataType).toArray
+
+      override def hasNext: Boolean = hostProducer.hasNext
+
+      /**
+       * At here, for consuming AsyncProducer effectively, we would like to achieve two things:
+       * 1. Does NOT acquire GpuSemaphore until AsyncProducer is ready. It excludes the scenario
+       * that the task thread which holds GpuSemaphore doing nothing but waiting.
+       *
+       * 2. Does NOT commit the consumption of next element until getting GpuSemaphore. It means
+       * that the AsyncProducerThread will not be awakened until the previous element starts to
+       * be transferred to device.
+       */
+      override def next(): ColumnarBatch = {
+        // 1. Preparing Stage
+        // Before waiting for asynchronous cpu task, releases the potential GpuSemaphore being
+        // held by this thread. It may happen if there exists downstreaming nodes which coalesces
+        // multiple input batches (such as Aggregate).
+        Option(TaskContext.get()).foreach { ctx =>
+          GpuSemaphore.releaseIfNecessary(ctx)
```

Comment:
I would love to see the impact to NDS for just this part of the change. I am not saying that this is a bad option. I just want to quantify this vs other changes in the code base.

Reply:
I got an impression that overall "preload" feature didn't make too much different with NDS in our environment. Basically, this feature is to improve overall performance for time-consuming scan. thoughts? @sperlingxx

Reply:
Finally, I got some available slots on testing the `Release or NOT` on NDS(spark-2a). I ran each condition multiple times. The result showed there is no significant difference between release or NOT on NDS. @revans2 @winningsix 


|                   | NoRelease | Release  |
|-----------|-----------|-----------|
| Preload(0)  | 689s  | 693s  |
| Preload(3)  | 695s  | 672s  |